### DISCUSSION TITLE
Compiling sportstracker on OS-X

### DISCUSSION THREAD

**Discussion by janlul** [ssaring/sportstracker#242] [Q&A] [answered]
Hi,

I compiled the sportstracker sources before although I am new to Java and Maven. Last time it kind of worked so I dared to build the last update. Got me the latest sources and updated Java to AdoptOpenJDK16, Maven was already at 3.6.3 and tried to build the sources which got me a list of error executing "mvn clean package" that I can't interpret and gives me no clue.
Anyone experiencing the same problem or knows what might be the cause?
The errors I get:
build@Mac-mini sportstracker-master % mvn clean package
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] leafletmap                                                         [jar]
[INFO] st-parent                                                          [pom]
[INFO] st-util                                                            [jar]
[INFO] st-exerciseviewer                                                  [jar]
[INFO] sportstracker                                                      [jar]
[INFO] 
[INFO] ------------------------< de.saring:leafletmap >------------------------
[INFO] Building leafletmap 1.0.6-SNAPSHOT                                 [1/5]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ leafletmap ---
[INFO] Deleting /Users/peter/Eigenwaarde/IntelliJIDEAProjects/sportstracker-master/leafletmap/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ leafletmap ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 26 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ leafletmap ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- kotlin-maven-plugin:1.4.31:compile (compile) @ leafletmap ---
[INFO] Applied plugin: 'kotlinx-serialization'
[ERROR] java.lang.ExceptionInInitializerError
	at com.intellij.pom.java.LanguageLevel.<clinit>(LanguageLevel.java:25)
	at com.intellij.core.CoreLanguageLevelProjectExtension.<init>(CoreLanguageLevelProjectExtension.java:26)
	at com.intellij.core.JavaCoreProjectEnvironment.<init>(JavaCoreProjectEnvironment.java:42)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreProjectEnvironment.<init>(KotlinCoreProjectEnvironment.kt:26)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$ProjectEnvironment.<init>(KotlinCoreEnvironment.kt:121)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.createForProduction(KotlinCoreEnvironment.kt:425)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.createCoreEnvironment(K2JVMCompiler.kt:226)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:152)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:88)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.execCompiler(KotlinCompileMojoBase.java:228)
	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execCompiler(K2JVMCompileMojo.java:237)
	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execCompiler(K2JVMCompileMojo.java:55)
	at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.execute(KotlinCompileMojoBase.java:209)
	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execute(K2JVMCompileMojo.java:222)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected void java.util.ResourceBundle.setParent(java.util.ResourceBundle) accessible: module java.base does not "opens java.util" to unnamed module @7209ffb5
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at com.intellij.util.ReflectionUtil.makeAccessible(ReflectionUtil.java:252)
	at com.intellij.util.ReflectionUtil.getDeclaredMethod(ReflectionUtil.java:269)
	at com.intellij.DynamicBundle.<clinit>(DynamicBundle.java:22)
	... 39 more

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] leafletmap 1.0.6-SNAPSHOT .......................... FAILURE [  4.098 s]
[INFO] st-parent 1.0.0 .................................... SKIPPED
[INFO] st-util 7.8.1-SNAPSHOT ............................. SKIPPED
[INFO] st-exerciseviewer 7.8.1-SNAPSHOT ................... SKIPPED
[INFO] sportstracker 7.8.1-SNAPSHOT ....................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.289 s
[INFO] Finished at: 2021-04-10T16:24:29+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.4.31:compile (compile) on project leafletmap: Compilation failure
[ERROR] java.lang.ExceptionInInitializerError
[ERROR] 	at com.intellij.pom.java.LanguageLevel.<clinit>(LanguageLevel.java:25)
[ERROR] 	at com.intellij.core.CoreLanguageLevelProjectExtension.<init>(CoreLanguageLevelProjectExtension.java:26)
[ERROR] 	at com.intellij.core.JavaCoreProjectEnvironment.<init>(JavaCoreProjectEnvironment.java:42)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreProjectEnvironment.<init>(KotlinCoreProjectEnvironment.kt:26)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$ProjectEnvironment.<init>(KotlinCoreEnvironment.kt:121)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.createForProduction(KotlinCoreEnvironment.kt:425)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.createCoreEnvironment(K2JVMCompiler.kt:226)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:152)
[ERROR] 	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
[ERROR] 	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:88)
[ERROR] 	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
[ERROR] 	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
[ERROR] 	at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.execCompiler(KotlinCompileMojoBase.java:228)
[ERROR] 	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execCompiler(K2JVMCompileMojo.java:237)
[ERROR] 	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execCompiler(K2JVMCompileMojo.java:55)
[ERROR] 	at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.execute(KotlinCompileMojoBase.java:209)
[ERROR] 	at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execute(K2JVMCompileMojo.java:222)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[ERROR] Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected void java.util.ResourceBundle.setParent(java.util.ResourceBundle) accessible: module java.base does not "opens java.util" to unnamed module @7209ffb5
[ERROR] 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
[ERROR] 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
[ERROR] 	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
[ERROR] 	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
[ERROR] 	at com.intellij.util.ReflectionUtil.makeAccessible(ReflectionUtil.java:252)
[ERROR] 	at com.intellij.util.ReflectionUtil.getDeclaredMethod(ReflectionUtil.java:269)
[ERROR] 	at com.intellij.DynamicBundle.<clinit>(DynamicBundle.java:22)
[ERROR] 	... 39 more
[ERROR] 
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
build@Mac-mini sportstracker-master % mvn --version    
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /Users/build/apache-maven-3.6.3
Java version: 16, vendor: AdoptOpenJDK, runtime: /Library/Java/JavaVirtualMachines/adoptopenjdk-16.jdk/Contents/Home
Default locale: nl_NL, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
build@Mac-mini sportstracker-master %

**Comment by ssaring** [comment, marked as answer]
This is a known Kotlin compatibility issue with JDK 16+, that's also commented in the [TODO.txt](https://github.com/ssaring/sportstracker/blob/master/sportstracker/docs/TODO.txt) 
Kotlin 1.5.0 will provide a proper solution, meanwhile you need to bypass this problem by setting the environment variable MAVEN_OPTS:

`export MAVEN_OPTS=--illegal-access=permit`

You can found further information in the TODO.txt.

**Reply by janlul** [reply]
Thanks that worked well. After adjusting the package script with the right name it could also be packaged

**Reply by ssaring** [reply]
Yep, the constants in the packaging scripts always needs to be adjusted, the commited version is from the latest release. Although that's documented, automation would be nicer ;)

**Comment by janlul** [comment]
Hi Stefan,

I have been fiddling a bit to make heart rate zones work. For now it is listing and showing them only but it looks to me as a start to integrate it with the existing usage.
Since I could not upload files I send them this way and you can stay in charge of your code.

I envision that there is a store of heart rate zones similar to weights which can be manually and automatically updated.
So anyone can use these whether or not the device provides this information.
When a device is able to provide heart rate zones the last in the list ought to be checked and if the devices provides different heart rates a new set is made. This will take some extra time, since I now concentrated on the first part, the list and did not dive into the existing use of the heart rate zones.
Of course I had to make some design decisions: 
- implemented 5 zones for it covers most of the heart rate zone models and did not make the number of zones optional for I am not fluent in Java
- treated it similar to weight, this enables storage and infrequent changes
- edited some icons, funny to do but it does not reflect the applications style yet

Hope you like the idea and could add parts of it in the future.

Kind regards

Peter

**Comment by janlul** [comment]
HI Stefan,

Worked a bit on the remaining issues and at least tackled the reading of heart rate zones from the FIT-file. Made the reading dimension independent, still have to figure out how to match the list and the reading. 
Issue not implemented yet is how to deal with the plethora of heart rate zone models although I guess it will be allowing for the maximum possible zones and only filling what is available. 

Update contains all files modified from 7.8.2.

Kind regards

Peter

**Reply by ssaring** [reply]
Hi Peter,

thanks for your input regarding the heartrate zones. I was already often thinking how to manage that and there were also multiple user requests and suggestions on this topic. 
Most of the ideas are based on adding a heartrate zones model defined by the user, this will be then applied to the displayed exercises. But that ideas does not convince me completely for several reasons:

1. Users can have multiple heartrate zones for each sport type (quite complicated)
2. The heartrate zones will change over the years, so it might be not valid anymore when e.g. displaying an exercise recorded 7 years ago. So a kind of history or archive would be needed (quite complicated)
3. Many watches can provide the heartrate zones in the recorded exercises, what is being preferred then?

With "complicated" I mean that this is not very user friendly, this needs a lot of description and knowledge. It's always a tradeoff, but a simplistic UI/UX is often more important than features for me. And yes, it's also more effort from the technical side.

Therefore I favor a simiplistic solution which tries to read and display the heartrate zones defined by the training devices. So when the ExerciseParser implementation is able to read them, then the ExerciseViewer can display them. I have not investigated the FIT file format that much in detail, but it seems that you were able to find the info. Unfortunately I can't see any attachment on this discussion with the sources.
If this can be implemented for FIT files, then this can be used for all the Garmin and Suunto users. This should be the majority (I'm using Garmin's).

Bye, Stefan

**Reply by janlul** [reply]
I see something went wrong in attaching the files, apparently .tgz is not supported. I 'll call the .tgz file .zip
[sport_tracker_with_hrz20210414.zip](https://github.com/ssaring/sportstracker/files/6309093/sport_tracker_with_hrz20210414.zip)
If the wish is to have heart rate zones per sport type this would introduce an extra dimension indeed. Given the setup I choose it might be addressed by introducing this parameter per set of zones along the date.

I recognize the time dependance, that is why I choose to use the list of heart rate zones. Since I am not fluent in Java/Kotlin I started with a fixed number of zones but I already found out how to make it flexible in the parser part so I will try to make the list implementation also flexible, will be somewhere next week.
There are also devices that do not provide the zones (Suunto Spartan is among those :)) in which case the list can help to provide the zones nonetheless. 
The approach I envision is to detect heart rate zones while parsing and if they do not exist yet (this is a tricky check I think) they are entered in the list. From there the list can be used to show heart rate zone on a given date (and sport type if implemented) or if not provided enter them manually.

**Comment by ssaring** [comment]
I've fiddled today a bit with the Garmin FIT parser implementation with the goal to parse the current user heartrate zones from the recorded activity files. Although the FIT format documentation contains such data, I was not able to read them from my recorded FIT files (Garmin Fenix 6), this was just missing in the appropriate message data.

**Reply by janlul** [reply]
That is a bummer. Can you find the data when you read the output of FitCSVTool.jar tool from the Garmin FIT SDK? That way I found out Suunto stores those values there.

**Comment by ssaring** [comment]
Interesting, by inspecting the converted CSV tool I was able to find a message record which contains the hr zone ranges, the max heartrate and the lactate threashold. Unfortunately it's not clear what kind of message and field this is, these are just listed as "unknown". But this is at least a good start :)

Thanks for the hint with the FitCSVTool.

**Reply by janlul** [reply]
Interesting filed " unknown" :) 
Happy you have a start to find the entries. 
I couldn't hold in so I am looking into the way how to make the zone list more flexible. For now I am extending the internals to 10 zones (should do for most models) and will have to look in how to limit what is shown.

**Reply by janlul** [reply]
Went into the various datasets of the FIT SDK and implemented a few of them to have a starting point to add relevant data on Main and Optional tabs. Also tackled the device recognition issue I had in this search
[sport_tracker_with_hrz20210416.tar.zip](https://github.com/ssaring/sportstracker/files/6327380/sport_tracker_with_hrz20210416.tar.zip)
. Attached the changed files compared to 7.8.2

**Comment by ssaring** [comment]
Finally I was able to parse the heartrate zone information from the Garmin FIT files, if it is available (is there on newer devices). You can find the details in the PullRequest https://github.com/ssaring/sportstracker/pull/244 which has already been merged to the master branch.

The majority of users should now be able to see the heartrate zone information directly from the recorded HRM files without the need for manual configuration and maintainance of these zones. And the displayed zones are now valid for the date of the displayed exercise. 

Therefore I would prefer not to add an additional user heartrate zone configuration, it just makes the application more complex. And it would also be not clear, which zone configuration is displayed when there is a user configuration and the availability in the HRM file.

**Reply by janlul** [reply]
Agree for the user entering and complexity. 
Would be nice though to have an overview of the heart rate zones over time like Weight. Since weight sometimes can be obtained from the FIT file these two can be treated the same if passed from ExerciseViewer to SportsTracker.

**Comment by ssaring** [comment]
Regarding the weight I'm a bit sceptical. Yes, it's stored in almost all FIT files, but this is just the weight the user enters once when configuring the initial user profile for the e.g. Garmin watch. This is typically not maintained and fixed for years. It might be maintained automatically when you have a digial scale connected to Garmin Connect, but I guess this will only be the case for a very small fraction of users.
